<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="utf-8">
    <title>JSDoc: Source: menu_genomegraph_2.js</title>

    <script src="scripts/prettify/prettify.js"> </script>
    <script src="scripts/prettify/lang-css.js"> </script>
    <!--[if lt IE 9]>
      <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->
    <link type="text/css" rel="stylesheet" href="styles/prettify-tomorrow.css">
    <link type="text/css" rel="stylesheet" href="styles/jsdoc-default.css">
</head>

<body>

<div id="main">

    <h1 class="page-title">Source: menu_genomegraph_2.js</h1>

    



    
    <section>
        <article>
            <pre class="prettyprint source linenums"><code>/**
 * view an experiment over a subfam
 */

function menu_genomegraph_2()
{
    /* view an experiment over a subfam
     */
    menu_hide();
    for(var key in apps.gg.view) {
        var v=apps.gg.view[key];
        if(v.type==2 &amp;&amp; v.geoid==gflag.menu.geoid &amp;&amp; v.subfamid==gflag.menu.subfamid) {
            view_gg(key);
            return;
        }
    }
    var geoinfo=id2geo[gflag.menu.geoid];
    var sfinfo=id2subfam[gflag.menu.subfamid];
    var has_consensus=sfinfo.consensuslen>0;
// TODO check if this view has been made

    /* generate this view */
    var viewobj={
        type:2,
        subfamid:gflag.menu.subfamid,
        geoid:gflag.menu.geoid,
        tabs:{info:[],cons:[],bev:[],rank:[]},
        bev:{},
        rank:{},
        has_input:geoinfo.input!=null,
        tklst:[],
    };
    var key=init_newgg(viewobj);

    viewobj.header.appendChild(make_geohandle(gflag.menu.geoid));
    dom_addtext(viewobj.header,'&amp;nbsp;&amp;nbsp;over&amp;nbsp;&amp;nbsp;');
    viewobj.header.appendChild(make_subfamhandle(gflag.menu.subfamid));

    /*******
     tab headers
     **/
    var b=dom_addbutt(viewobj.header,'Description',tab_click);
    b.which=1;
    b.key=key;
    b.style.marginLeft=20;
    b.disabled=true;
    viewobj.tabs.info[0]=b;
    if(has_consensus) {
        b=dom_addbutt(viewobj.header,'Loading...',tab_click);
        b.which=2;
        b.key=key;
        viewobj.tabs.cons[0]=b;
    } else {
        viewobj.tabs.cons=[document.createElement('button'),document.createElement('div')];
    }
// bev
    b=dom_addbutt(viewobj.header,'Loading...',tab_click);
    b.key=key;
    b.which=3;
    viewobj.tabs.bev[0]=b;

    /**********
     first section: geo/tk info, consensus wiggle plot
     */
    var table=dom_create('table',viewobj.content);
    viewobj.tabs.info[1]=table;
// row 1
    var tr=table.insertRow(0);
    var td=tr.insertCell(0);
    td.className='tph';
    td.innerHTML='Experiment';
    tr.insertCell(1).appendChild(make_geohandle(gflag.menu.geoid));
// row 2
    tr=table.insertRow(-1);
    td=tr.insertCell(0);
    td.className='tph';
    td.innerHTML=viewobj.has_input==null?'Data files':'Treatment data files';
    td=tr.insertCell(1);
    for(var i=0; i&lt;geoinfo.treatment.length; i++) {
        td.appendChild(make_filehandle(geoinfo.treatment[i], viewobj.has_input?'treatment':'file',geoinfo.treatment.length,i));
    }
// row 2.5
    if(viewobj.has_input) {
        tr=table.insertRow(-1);
        td=tr.insertCell(0);
        td.className='tph';
        td.innerHTML='Input/Control data files';
        td=tr.insertCell(1);
        for(var i=0; i&lt;geoinfo.input.length; i++) {
            td.appendChild(make_filehandle(geoinfo.input[i], 'input', geoinfo.input.length,i));
        }
    }
// row 3
    tr=table.insertRow(-1);
    td=tr.insertCell(0);
    td.className='tph';
    td.innerHTML='Log ratio of average score';
    tr.insertCell(1).innerHTML=gflag.menu.pointdata[0]+' &lt;span style="font-size:10px;">iteres&lt;/span>&lt;br>'+gflag.menu.pointdata[1]+' &lt;span style="font-size:10px;">BWA&lt;/span>';
// row 4
    tr=table.insertRow(-1);
    td=tr.insertCell(0);
    td.className='tph';
    td.innerHTML='Repeat subfamily';
    tr.insertCell(1).appendChild(make_subfamhandle(viewobj.subfamid));

    /*******
     second section, consensus plot
     TODO draw for all subfams
     */
    if(sfinfo.consensuslen>0) {
        var d=dom_create('div',viewobj.content);
        d.style.display='none';
        viewobj.tabs.cons[1]=d;
        var d2=document.createElement('div');
        d.appendChild(d2);
        if(viewobj.has_input) {
            var s=document.createElement('span');
            s.innerHTML='Treatment read count&amp;nbsp;';
            d2.appendChild(s);
            s=document.createElement('span');
            s.style.backgroundColor='rgb('+qtc_treat_a.pr+','+qtc_treat_a.pg+','+qtc_treat_a.pb+')';
            s.style.color='black';
            s.innerHTML='&amp;nbsp;iteres&amp;nbsp;';
            d2.appendChild(s);
            s=document.createElement('span');
            s.style.backgroundColor='rgb('+qtc_treat_u.pr+','+qtc_treat_u.pg+','+qtc_treat_u.pb+')';
            s.style.color='black';
            s.innerHTML='&amp;nbsp;BWA&amp;nbsp;';
            d2.appendChild(s);
            s=document.createElement('span');
            s.innerHTML='&amp;nbsp;&amp;nbsp;Input read count&amp;nbsp;';
            d2.appendChild(s);
            s=document.createElement('span');
            s.style.backgroundColor='rgb('+qtc_input_a.pr+','+qtc_input_a.pg+','+qtc_input_a.pb+')';
            s.style.color='black';
            s.innerHTML='&amp;nbsp;iteres&amp;nbsp;';
            d2.appendChild(s);
            s=document.createElement('span');
            s.style.backgroundColor='rgb('+qtc_input_u.pr+','+qtc_input_u.pg+','+qtc_input_u.pb+')';
            s.style.color='black';
            s.innerHTML='&amp;nbsp;BWA&amp;nbsp;';
            d2.appendChild(s);
        } else {
            var s=document.createElement('span');
            s.innerHTML='All aligned reads by iteres&amp;nbsp;';
            d2.appendChild(s);
            s=document.createElement('span');
            s.style.backgroundColor='rgb('+qtc_treat_a.pr+','+qtc_treat_a.pg+','+qtc_treat_a.pb+')';
            s.style.color='black';
            s.innerHTML='&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;';
            d2.appendChild(s);
            s=document.createElement('span');
            s.innerHTML='Uniquely aligned reads by BWA&amp;nbsp;';
            d2.appendChild(s);
            s=document.createElement('span');
            s.style.backgroundColor='rgb('+qtc_treat_u.pr+','+qtc_treat_u.pg+','+qtc_treat_u.pb+')';
            s.style.color='black';
            s.innerHTML='&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;';
            d2.appendChild(s);
        }
        var s=document.createElement('span');
        s.innerHTML='&amp;#10092;ALL&amp;#10093;';
        s.className='clb4';
        s.style.display='none';
        s.style.marginLeft=50;
        s.key=key;
        s.addEventListener('click', consensusPlot_showall, false);
        viewobj.consensusplot={
            allbutt:s,
            scrollable:[],
            preset_left:0,
            tk2canvas:{},
        };
        d2.appendChild(s);

        d2=document.createElement('div');
        d.appendChild(d2);
        viewobj.consensuswigglediv=d2;
        browser.ajax('repeatbrowser=on&amp;getconsensuswig=on&amp;rpbrDbname='+infodb+'&amp;geo='+geoinfo.acc+'&amp;subfam='+sfinfo.name+'&amp;viewkey='+key,function(data){
            /* construct canvas and scroll holders
             */
            var vobj=apps.gg.view[data.key];
            var holder=vobj.consensuswigglediv;
            var cplot=vobj.consensusplot;

            var conlen=id2subfam[vobj.subfamid].consensuslen;
            cplot.holderwidth=800;
            cplot.sf=cplot.holderwidth/conlen; // px per bp

            var table=document.createElement('table');
            table.style.marginTop=20;
            holder.appendChild(table);

            /* bp ruler */
            var tr=table.insertRow(0);
            tr.insertCell(0);
            var td=tr.insertCell(1);
            var div=document.createElement('div');
            div.className='scholder';
            div.style.width=cplot.holderwidth;
            td.appendChild(div);
            var c=document.createElement('canvas');
            c.key=data.key;
            c.className='sclb';
            c.height=div.style.height=15; // hard coded?
            c.addEventListener('mousedown',consensusPlot_ruler_md,false);
            div.appendChild(c);
            cplot.scrollable.push(c);
            cplot.rulercanvas=c;

            /* get max/min from assay tracks */
            var max=0, min=0;
            var tmp=wiggle_maxmin(data.treatment_all);
            if(max&lt;tmp[0]) max=tmp[0];
            if(min>tmp[1]) min=tmp[1];
            tmp=wiggle_maxmin(data.treatment_unique);
            if(max&lt;tmp[0]) max=tmp[0];
            if(min>tmp[1]) min=tmp[1];
            if(vobj.has_input) {
                tmp=wiggle_maxmin(data.input_all);
                if(max&lt;tmp[0]) max=tmp[0];
                if(min>tmp[1]) min=tmp[1];
                tmp=wiggle_maxmin(data.input_unique);
                if(max&lt;tmp[0]) max=tmp[0];
                if(min>tmp[1]) min=tmp[1];
            }
            cplot.assayMax=max;
            cplot.assayMin=min;

            /* wiggle treat-all */
            for(var i=0; i&lt;data.treatment_all.length; i++) {
                var dd=data.treatment_all[i];
                var tr=table.insertRow(-1);
                var td=tr.insertCell(0);
                td.align='right';
                td.vAlign='middle';
                td.appendChild(make_filehandle(dd[0], vobj.has_input?'treatment':'file', id2geo[vobj.geoid].treatment.length, i));

                td=tr.insertCell(1);
                div=document.createElement('div');
                div.className='scholder';
                div.style.width=cplot.holderwidth;
                td.appendChild(div);

                var c=document.createElement('canvas');
                c.className='sclb';
                c.key=data.key;
                c.addEventListener('mousedown', consensusPlot_md,false);
                div.appendChild(c);
                c.height=div.style.height=wiggleheight+densitydecorpaddingtop;
                cplot.tk2canvas[dd[0]]=c;
                cplot.scrollable.push(c);
            }
            cplot.treatment_all=data.treatment_all;
            cplot.treatment_unique=data.treatment_unique;

            if(vobj.has_input) {
                // plot input-all
                for(i=0; i&lt;data.input_all.length; i++) {
                    var dd=data.input_all[i];
                    var tr=table.insertRow(-1);
                    var td=tr.insertCell(0);
                    td.align='right';
                    td.vAlign='middle';
                    td.appendChild(make_filehandle(dd[0], 'input', id2geo[vobj.geoid].input.length, i));
                    td=tr.insertCell(1);
                    div=document.createElement('div');
                    div.className='scholder';
                    div.style.width=cplot.holderwidth;
                    td.appendChild(div);

                    var c=document.createElement('canvas');
                    c.className='sclb';
                    c.key=data.key;
                    c.addEventListener('mousedown', consensusPlot_md,false);
                    div.appendChild(c);
                    c.height=div.style.height=wiggleheight+densitydecorpaddingtop;
                    cplot.tk2canvas[dd[0]]=c;
                    cplot.scrollable.push(c);
                }
                cplot.input_all=data.input_all;
                cplot.input_unique=data.input_unique;
            }

            /* density wiggle */
            var dmax=0;
            for(i=0; i&lt;data.density.length; i++) {
                var v=data.density[i];
                if(dmax&lt;v) dmax=v;
            }
            cplot.densityMax=dmax;
            cplot.densitydata=data.density;

            tr=table.insertRow(-1);
            td=tr.insertCell(0);
            td.align='center';
            td.vAlign='middle';
            td.innerHTML='TE coverage';
            td=tr.insertCell(1);
            div=document.createElement('div');
            div.className='scholder';
            div.style.width=cplot.holderwidth;
            td.appendChild(div);
            c=document.createElement('canvas');
            c.className='sclb';
            c.key=data.key;
            c.addEventListener('mousedown', consensusPlot_md,false);
            div.appendChild(c);
            c.height=div.style.height=wiggleheight+densitydecorpaddingtop;
            cplot.scrollable.push(c);
            cplot.densitycanvas=c;

            /* sequence */
            if('consensusseq' in data) {
                cplot.sequence=data.consensusseq;
                tr=table.insertRow(-1);
                td=tr.insertCell(0);
                td.align='center';
                td.vAlign='middle';
                td.style.fontWeight='bold';
                td.innerHTML= '&lt;span style="color:'+ntbcolor.a+'">A&lt;/span>'+
                    '&lt;span style="color:'+ntbcolor.t+'">T&lt;/span>'+
                    '&lt;span style="color:'+ntbcolor.c+'">C&lt;/span>'+
                    '&lt;span style="color:'+ntbcolor.g+'">G&lt;/span>'+
                    '&lt;span style="color:'+ntbcolor.n+'">N&lt;/span>';
                td=tr.insertCell(1);
                div=document.createElement('div');
                div.className='scholder';
                div.style.width=cplot.holderwidth;
                td.appendChild(div);
                c=document.createElement('canvas');
                c.className='sclb';
                c.key=data.key;
                c.addEventListener('mousedown', consensusPlot_md,false);
                div.appendChild(c);
                c.height=div.style.height=ideoHeight;
                cplot.scrollable.push(c);
                cplot.seqcanvas=c;
            }

            vobj.tabs.cons[0].innerHTML='Consensus';

            draw_consensusPlot(data.key);
        });
    }

    /********************
     third section, bev
     */
    var d=dom_create('div',viewobj.content);
    d.style.display='none';
    viewobj.tabs.bev[1]=d;
    viewobj.bev.holder=d;
    viewobj.bev.scoreType=viewobj.has_input?'Log ratio':'Track score';
    viewobj.bev.ismain=true;
    viewobj.colorscale={};
// this part is duplicated in tkentryclick_add2gg
    browser.ajax('repeatbrowser=on&amp;getsubfamcopieswithtk=on'+subfamtrackparam(viewobj.subfamid)+'&amp;geo='+geoinfo.acc+'&amp;viewkey='+key,function(data){
        var vobj=apps.gg.view[data.key];
        make_genomebev_base(vobj,data.key);

        parseData_exp_bev(data,vobj,vobj.bev);

        make_bevcolorscale(vobj,vobj.bev,data.key);

        draw_genomebev_experiment(vobj,vobj.bev);

        vobj.tabs.bev[0].innerHTML='Genome';

        /* rank the repeats by assay value
         */
        var tosort=[];
        for(var chr in vobj.bev.data) {
            var lst=vobj.bev.data[chr];
            for(i=0; i&lt;lst.length; i++)
                tosort.push([chr, i, lst[i][4]]);
            /* 0: chr, 1: in-chr array idx, 2: data
             */
        }
        tosort.sort(scoreSort);
        var _r=vobj.rank;
        _r.rarr=tosort;

    });

    simulateEvent(viewobj.tabs.info[0],'click');

    view_gg(key);
    cloakPage();
    panelFadein(apps.gg.main,50,20);
    document.getElementById('viewlstbutt').style.display='inline';
    var c=document.getElementById('viewcount');
    c.innerHTML=parseInt(c.innerHTML)+1;
    /* over, geo+subfam */
}</code></pre>
        </article>
    </section>




</div>

<nav>
    <h2><a href="index.html">Home</a></h2><h3>Global</h3><ul><li><a href="global.html#add2gg_invoketkselect">add2gg_invoketkselect</a></li><li><a href="global.html#afteraddremoveview">afteraddremoveview</a></li><li><a href="global.html#alethiometer_addtk_cb">alethiometer_addtk_cb</a></li><li><a href="global.html#applyweight_do">applyweight_do</a></li><li><a href="global.html#applyweight_tip">applyweight_tip</a></li><li><a href="global.html#beam_rankitem">beam_rankitem</a></li><li><a href="global.html#bigmap_move_Md">bigmap_move_Md</a></li><li><a href="global.html#bigmap_move_Mm">bigmap_move_Mm</a></li><li><a href="global.html#bigmap_move_Mu">bigmap_move_Mu</a></li><li><a href="global.html#browser">browser</a></li><li><a href="global.html#click_gghandle">click_gghandle</a></li><li><a href="global.html#colheader_getlanding">colheader_getlanding</a></li><li><a href="global.html#colheader_Mmove">colheader_Mmove</a></li><li><a href="global.html#colheader_Mout">colheader_Mout</a></li><li><a href="global.html#colheader_select_Md">colheader_select_Md</a></li><li><a href="global.html#colheader_select_Mm">colheader_select_Mm</a></li><li><a href="global.html#colheader_select_Mu">colheader_select_Mu</a></li><li><a href="global.html#colorscale_slider_md">colorscale_slider_md</a></li><li><a href="global.html#colorscale_slider_mm">colorscale_slider_mm</a></li><li><a href="global.html#colorscale_slider_mu">colorscale_slider_mu</a></li><li><a href="global.html#colorscale_slidermoved">colorscale_slidermoved</a></li><li><a href="global.html#consensusPlot_md">consensusPlot_md</a></li><li><a href="global.html#consensusPlot_mm">consensusPlot_mm</a></li><li><a href="global.html#consensusPlot_mu">consensusPlot_mu</a></li><li><a href="global.html#consensusPlot_ruler_md">consensusPlot_ruler_md</a></li><li><a href="global.html#consensusPlot_ruler_mm">consensusPlot_ruler_mm</a></li><li><a href="global.html#consensusPlot_ruler_mu">consensusPlot_ruler_mu</a></li><li><a href="global.html#consensusPlot_showall">consensusPlot_showall</a></li><li><a href="global.html#coordSort">coordSort</a></li><li><a href="global.html#delete_bev_experiment">delete_bev_experiment</a></li><li><a href="global.html#delete_gg">delete_gg</a></li><li><a href="global.html#delete_view">delete_view</a></li><li><a href="global.html#draw_consensusPlot">draw_consensusPlot</a></li><li><a href="global.html#draw_genomebev_1">draw_genomebev_1</a></li><li><a href="global.html#draw_genomebev_experiment">draw_genomebev_experiment</a></li><li><a href="global.html#drawBigmap">drawBigmap</a></li><li><a href="global.html#filehandle_click">filehandle_click</a></li><li><a href="global.html#find_te">find_te</a></li><li><a href="global.html#find_te_ku">find_te_ku</a></li><li><a href="global.html#genomebev_splinter_mm">genomebev_splinter_mm</a></li><li><a href="global.html#genomebev_splinter_mu">genomebev_splinter_mu</a></li><li><a href="global.html#genomebev_tooltip_mousemove">genomebev_tooltip_mousemove</a></li><li><a href="global.html#genomebev_zoomin_md">genomebev_zoomin_md</a></li><li><a href="global.html#geohandle_click">geohandle_click</a></li><li><a href="global.html#get_file_info">get_file_info</a></li><li><a href="global.html#htmltext_bigmapcell">htmltext_bigmapcell</a></li><li><a href="global.html#htmltext_subfaminfo">htmltext_subfaminfo</a></li><li><a href="global.html#init_newgg">init_newgg</a></li><li><a href="global.html#make_a_track">make_a_track</a></li><li><a href="global.html#make_bevcolorscale">make_bevcolorscale</a></li><li><a href="global.html#make_filehandle">make_filehandle</a></li><li><a href="global.html#make_genomebev_base">make_genomebev_base</a></li><li><a href="global.html#make_geohandle">make_geohandle</a></li><li><a href="global.html#make_subfamhandle">make_subfamhandle</a></li><li><a href="global.html#menu_bigmap">menu_bigmap</a></li><li><a href="global.html#menu_click4geo">menu_click4geo</a></li><li><a href="global.html#menu_genomegraph_1">menu_genomegraph_1</a></li><li><a href="global.html#menu_genomegraph_2">menu_genomegraph_2</a></li><li><a href="global.html#menu_getExperimentInfo">menu_getExperimentInfo</a></li><li><a href="global.html#menu_getExperimentInfo_2">menu_getExperimentInfo_2</a></li><li><a href="global.html#menu_showsearchui">menu_showsearchui</a></li><li><a href="global.html#menu_temcm">menu_temcm</a></li><li><a href="global.html#pane_hide">pane_hide</a></li><li><a href="global.html#parseData_exp_bev">parseData_exp_bev</a></li><li><a href="global.html#readygo">readygo</a></li><li><a href="global.html#repeatbrowser_load">repeatbrowser_load</a></li><li><a href="global.html#repeatbrowser_loadhub_recursive">repeatbrowser_loadhub_recursive</a></li><li><a href="global.html#repeatbrowser_loadhub_recursive_cb">repeatbrowser_loadhub_recursive_cb</a></li><li><a href="global.html#scatterplot_clickmenu_2">scatterplot_clickmenu_2</a></li><li><a href="global.html#scatterplot_dotclick_2">scatterplot_dotclick_2</a></li><li><a href="global.html#scatterplot_show_2">scatterplot_show_2</a></li><li><a href="global.html#scatterplot_submit2">scatterplot_submit2</a></li><li><a href="global.html#scoredistribution_mm">scoredistribution_mm</a></li><li><a href="global.html#scoreSort">scoreSort</a></li><li><a href="global.html#show_info">show_info</a></li><li><a href="global.html#show_viewlst">show_viewlst</a></li><li><a href="global.html#sortcolumn_bytrack">sortcolumn_bytrack</a></li><li><a href="global.html#subfamhandle_click">subfamhandle_click</a></li><li><a href="global.html#subfamtrackparam">subfamtrackparam</a></li><li><a href="global.html#tab_click">tab_click</a></li><li><a href="global.html#temcm_click">temcm_click</a></li><li><a href="global.html#tkentryclick_add2gg">tkentryclick_add2gg</a></li><li><a href="global.html#view_gg">view_gg</a></li><li><a href="global.html#wiggle_maxmin">wiggle_maxmin</a></li><li><a href="global.html#zoomout">zoomout</a></li></ul>
</nav>

<br class="clear">

<footer>
    Documentation generated by <a href="https://github.com/jsdoc3/jsdoc">JSDoc 3.3.2</a> on Fri Jul 24 2015 17:11:57 GMT-0500 (CDT)
</footer>

<script> prettyPrint(); </script>
<script src="scripts/linenumber.js"> </script>
</body>
</html>
